Make `--deny=...` support `added`, `changed`, and `removed`

matthiasbeyer opened the pull request:
In #72 the foundation for denying API diffs was implemented.

This PR is the next step to allow only certain changes to APIs via CLI parameter.

Tell me what you think, whether this looks good to you. If it does, I can of course add tests!

Enselic commented:
Thank you for making a follow-up PR!

You have real bad luck with CI because you yet again hit a failure not caused by your changes.

If you rebase on main branch after fetch it should pass.

As for the PR itself, some high level comments:
* The terminology we should use is [added, changed, removed](https://docs.rs/public-api/latest/public_api/diff/struct.PublicItemsDiff.html) (not deleted). I might have wrongly use "deleted" myself in places, but it should be "removed".
* I'm thinking the CLI should look like this: `--deny=added` `--deny=removed` `--deny=changed`. Maybe even allow combinations such like this: `--deny=changed,removed`, but feel free to wait with that.
* Before a CR I would like tests to be in place and pass in CI, because that indicates significant maturity of the code, and I prefer to review mature code :)

matthiasbeyer commented:
Rebased, fixed your suggestions and also added tests. I'm not 100% sure about whether the tests are correct though, so you should probably have a look at them first :laughing:

Enselic commented:
Just a tip in case you didn't know: You can run `./scripts/run-ci-locally.sh` to see if CI will pass. 99% of the time the script is accurate. I would say 100% for code like in this PR. So you don't have to push code here to see if CI will pass :)

Enselic commented on cargo-public-api/tests/cargo-public-api-bin-tests.rs:
```diff
@@ -123,6 +173,60 @@ fn deny_with_diff() {
+    let mut cmd = Command::cargo_bin("cargo-public-api").unwrap();
+    cmd.current_dir(test_crate_path());
+    cmd.arg("--diff-git-checkouts");
+    cmd.arg("v0.0.4");
```
Heads-up: https://github.com/Enselic/cargo-public-api/pull/88 changes the names of tags, so you will want to use the new tag names

Enselic replied:
https://github.com/Enselic/cargo-public-api/pull/89 changes how tests are written so watch out for that one too

I think it will be straightforward to sort it out though.

matthiasbeyer replied:
Thanks for the heads-up!

Enselic commented:
Here's another heads-up: I just merged https://github.com/Enselic/cargo-public-api/pull/90.

It does not cause any merge conflict with this PR (I did a test-merge), but I recommend you rebase on main again to not accidentally start introducing conflicts in your own code.

matthiasbeyer commented:
Rebase was a cake :cake: and I took the liberty to squash away my fixups while I was at it.

Do you have any pointers what I need to fix to get the CI green? I am not seeing it right now :see_no_evil:

Enselic commented on cargo-public-api/tests/cargo-public-api-bin-tests.rs:
```diff
@@ -176,14 +229,65 @@ fn deny_with_diff() {
+    let mut cmd = Command::cargo_bin("cargo-public-api").unwrap();
+    cmd.current_dir(&test_repo.path);
+    cmd.arg("--diff-git-checkouts");
+    cmd.arg("v0.2.0");
```
The api diff between 0.2.0 and 0.3.0 is only removal of an item. Try diff between 0.1.0 and 0.2.0 instead, that changes and adds items.

Enselic commented on cargo-public-api/tests/cargo-public-api-bin-tests.rs:
```diff
@@ -176,14 +229,65 @@ fn deny_with_diff() {
+    cmd.assert()
+        .stderr(contains(
+            r#"Changed items in the public API
+===============================
```
Just took a quick glance and the test expects no changed items, but there are changed items. The test failure output tells you this :)

Enselic commented:
(The PR title is used in the autogenerated release-notes, so I changed it to be more descriptive)

Enselic commented on cargo-public-api/src/error.rs:
```diff
@@ -2,4 +2,13 @@
 pub enum Error {
     #[error("The API diff is not allowed as per --deny")]
     DiffDenied,
+
```
I've started to think some more about how we can/should structure this. I'm not sure what level of sophistication you are interested in, so if you think my demands are too high, let me know, and we can try to find a middle ground :)

But, the problem with this structure, is that it does not allow to represent the case of "items were both added and changed".

I think it would be better to have something akin to this structure. Let me know what you think about this:

```rust
use std::fmt::Display;

use public_api::PublicItem;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("The API diff is not allowed as per --deny: {0}")]
    DiffDenied(Violations),
}

#[derive(Debug)]
struct Violations {
    /// These items were added to the API, but no items may be added to the API.
    pub added: Vec<PublicItem>,

    /// These items were changed in the API, but no items may be changed in the API.
    pub changed: Vec<PublicItem>,

    /// These items were removed from the API, but no items may be removed from the API.
    pub removed: Vec<PublicItem>,
}

impl Display for Violations {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if !self.added.is_empty() {
            write!(f, "Added items not allowed: {:?} ", self.added)?
        }

        if !self.changed.is_empty() {
            write!(f, "Changed items not allowed: {:?} ", self.added)?
        }

        if !self.removed.is_empty() {
            write!(f, "Removed items not allowed: {:?} ", self.added)?
        }

        Ok(())
    }
}
```

NOTE: Not verified, just a sketch to get the point across.

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -130,12 +130,36 @@ fn main_() -> Result<()> {
+                match d {
+                    DenyMethod::All => {
+                        if !diff.is_empty() {
+                            anyhow::bail!(error::Error::DiffDenied)
```
Now you are using `bail!` again :)

I'd like the error to be able to capture multiple violations, and also write the list of what items that are causing the violation. (See other comment)

Enselic commented on cargo-public-api/tests/cargo-public-api-bin-tests.rs:
```diff
@@ -143,13 +143,66 @@ fn diff_public_items_with_dirty_tree_fails() {
 
 #[test]
 fn deny_when_not_diffing() {
+    let test_repo = TestRepo::new();
```
Did you get an error unless you did this? It might be needed, but so far I have not seen an error. But it could be because I'm lucky

matthiasbeyer replied:
I am not sure, I just added it in all tests... :-) didn't test without it.

Enselic commented on cargo-public-api/tests/cargo-public-api-bin-tests.rs:
```diff
@@ -176,14 +229,78 @@ fn deny_with_diff() {
     let mut cmd = Command::cargo_bin("cargo-public-api").unwrap();
     cmd.current_dir(&test_repo.path);
     cmd.arg("--diff-git-checkouts");
-    cmd.arg("v0.0.4");
```
Good move to adjust these 👍

Enselic commented:
I just saw that CI passes! Nice job! Maybe we should land this PR as-is (after review), and then I could look into doing my suggested improvements on top of that (but you are also more than welcomed to take a shot if you are interested)

Let me know your preference

matthiasbeyer commented:
I am a big fan of "get it right the first time", so I am happy to continue iterating on this!

matthiasbeyer commented on cargo-public-api/src/error.rs:
```diff
@@ -1,5 +1,61 @@
+    removed: Vec<PublicItem>,
+}
+
+impl Violations {
```
I briefly thought about adding a

```rust
pub fn into_error(self) -> Result<Error>
```

here, but then this might be a bit of an anti-pattern. Maybe you think otherwise?

matthiasbeyer commented on cargo-public-api/src/main.rs:
```diff
@@ -130,11 +130,48 @@ fn main_() -> Result<()> {
+            let mut violations = crate::error::Violations::new();
+            for d in deny {
+                match d {
+                    DenyMethod::All => {
+                        if !diff.added.is_empty() {
+                            violations.extend_added(diff.added.iter().cloned());
+                        }
+                        if !diff.changed.is_empty() {
+                            violations.extend_changed(diff.changed.iter().cloned());
+                        }
+                        if !diff.removed.is_empty() {
+                            violations.extend_removed(diff.removed.iter().cloned());
+                        }
+                    }
+
+                    DenyMethod::Added => {
+                        if !diff.added.is_empty() {
+                            violations.extend_added(diff.added.iter().cloned());
+                        }
+                    }
+
+                    DenyMethod::Changed => {
+                        if !diff.changed.is_empty() {
+                            violations.extend_changed(diff.changed.iter().cloned());
+                        }
+                    }
+
+                    DenyMethod::Removed => {
+                        if !diff.removed.is_empty() {
+                            violations.extend_removed(diff.removed.iter().cloned());
+                        }
+                    }
```
I really don't like the repetition here, do you have an idea how to make this nicer?

matthiasbeyer commented on cargo-public-api/src/main.rs:
```diff
@@ -130,11 +130,48 @@ fn main_() -> Result<()> {
+            }
+
+            if violations.is_empty() {
                 Ok(())
             } else {
-                Err(anyhow!(error::Error::DiffDenied))
+                Err(anyhow::Error::from(crate::error::Error::DiffDenied(
+                    violations,
+                )))
```
Bit clumsy.

What I could think of would be `violations.into_error()`, but that would make the `Violations` API ugly and sounds a bit like an anti-pattern.

Enselic commented on public-api/src/diff.rs:
```diff
@@ -106,12 +106,6 @@ impl PublicItemsDiff {
-
-    /// Check whether the diff is empty
-    #[must_use]
-    pub fn is_empty(&self) -> bool {
```
Heh, short-lived API 😅 

Note to self: Be careful when doing next version bump since this is a semver breaking change

Enselic replied:
Tests are failing because we remove this function. So let's keep this function instead of removing it. We needed it once, and we or others might need it later, and it is properly tested and everything. And seems to be low-risk to cause problems in the future.

Enselic commented on cargo-public-api/src/main.rs:
```diff
@@ -130,11 +130,48 @@ fn main_() -> Result<()> {
+        (Some(deny), Some(diff)) => {
+            let mut violations = crate::error::Violations::new();
+            for d in deny {
+                match d {
+                    DenyMethod::All => {
+                        if !diff.added.is_empty() {
+                            violations.extend_added(diff.added.iter().cloned());
+                        }
+                        if !diff.changed.is_empty() {
+                            violations.extend_changed(diff.changed.iter().cloned());
+                        }
+                        if !diff.removed.is_empty() {
+                            violations.extend_removed(diff.removed.iter().cloned());
+                        }
+                    }
+
+                    DenyMethod::Added => {
+                        if !diff.added.is_empty() {
+                            violations.extend_added(diff.added.iter().cloned());
+                        }
+                    }
+
+                    DenyMethod::Changed => {
+                        if !diff.changed.is_empty() {
+                            violations.extend_changed(diff.changed.iter().cloned());
+                        }
+                    }
+
+                    DenyMethod::Removed => {
+                        if !diff.removed.is_empty() {
+                            violations.extend_removed(diff.removed.iter().cloned());
+                        }
+                    }
+                }
```
Here is how I would simplify this:

```suggestion
                let deny_added = matches!(d, DenyMethod::Added | DenyMethod::All);
                let deny_changed = matches!(d, DenyMethod::Changed | DenyMethod::All);
                let deny_removed = matches!(d, DenyMethod::Removed | DenyMethod::All);
                if deny_added && !diff.added.is_empty() {
                    violations.extend_added(diff.added.iter().cloned());
                }
                if deny_changed && !diff.changed.is_empty() {
                    violations.extend_changed(diff.changed.iter().cloned());
                }
                if deny_removed && !diff.removed.is_empty() {
                    violations.extend_removed(diff.removed.iter().cloned());
                }
```

matthiasbeyer replied:
We could even further simplify this by adding functions on `DenyMethod` for indicating... 

```rust
pub fn is_deny_added(&self) -> bool {
    std::matches(self, DenyMethod::Added | DenyMethod::All)
}
```

and use them.

Enselic replied:
Sounds great!

Enselic reviewed (changes requested):
Thanks for your continued work on this! I did another review round, this time quite detailed.

matthiasbeyer commented on cargo-public-api/src/arg_types.rs:
```diff
@@ -9,6 +9,15 @@ use crate::{markdown::Markdown, output_formatter::OutputFormatter, plain::Plain}
+    /// Deny added things in API diffs
+    Added,
+
+    /// Deny changed things in API diffs
```
> Although I guess there is a risk with confusion since the tool output also has "changed"...

Yes I see that danger too!

Enselic commented on cargo-public-api/src/error.rs:
```diff
@@ -1,5 +1,63 @@
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.added.is_empty() || self.changed.is_empty() || self.removed.is_empty()
```
I tried this:

```bash
./scripts/create-test-git-repo.sh /tmp/foo
cargo run -- --manifest-path /tmp/foo/Cargo.toml --diff-git-checkouts v0.1.0 v0.2.0 --deny=added
```

and noticed it didn't fail. Debugged it a bit and noticed this is wrong, it should be like this:

```suggestion
        self.added.is_empty() && self.changed.is_empty() && self.removed.is_empty()
```

:)

matthiasbeyer replied:
Meh, thanks for catching that!

matthiasbeyer commented:
WHOOOP!

CI green! Can I ask you to spend a little portion of your Sunday for a review? If everything is fine, I will do a full history cleanup and then we can do a final round before we merge this - or, of course, if you find anything that can be improved, reiterate (I'm of course more than happy to)!

Enselic commented on cargo-public-api/src/arg_types.rs:
```diff
@@ -9,6 +9,29 @@ use crate::{markdown::Markdown, output_formatter::OutputFormatter, plain::Plain}
+}
+
+impl DenyMethod {
+    pub(crate) fn deny_added(&self) -> bool {
```
Looks like 1.63 finds some new lints, so let's fix these:

```
error: this argument (1 byte) is passed by reference, but would be more efficient if passed by value (limit: 8 byte)
  --> cargo-public-api/src/arg_types.rs:24:30
   |
24 |     pub(crate) fn deny_added(&self) -> bool {
   |                              ^^^^^ help: consider passing by value instead: `self`
   |
   = note: `-D clippy::trivially-copy-pass-by-ref` implied by `-D clippy::pedantic`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref

error: this argument (1 byte) is passed by reference, but would be more efficient if passed by value (limit: 8 byte)
  --> cargo-public-api/src/arg_types.rs:28:32
   |
28 |     pub(crate) fn deny_changed(&self) -> bool {
   |                                ^^^^^ help: consider passing by value instead: `self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref

error: this argument (1 byte) is passed by reference, but would be more efficient if passed by value (limit: 8 byte)
  --> cargo-public-api/src/arg_types.rs:32:32
   |
32 |     pub(crate) fn deny_removed(&self) -> bool {
   |                                ^^^^^ help: consider passing by value instead: `self`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref

error: consider adding a `;` to the last statement for consistent formatting
  --> cargo-public-api/src/error.rs:50:13
   |
50 |             write!(f, "Added items not allowed: {:?} ", self.added)?
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add a `;` here: `write!(f, "Added items not allowed: {:?} ", self.added)?;`
   |
   = note: `-D clippy::semicolon-if-nothing-returned` implied by `-D clippy::pedantic`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned

error: consider adding a `;` to the last statement for consistent formatting
  --> cargo-public-api/src/error.rs:54:13
   |
54 |             write!(f, "Changed items not allowed: {:?} ", self.changed)?
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add a `;` here: `write!(f, "Changed items not allowed: {:?} ", self.changed)?;`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned

error: consider adding a `;` to the last statement for consistent formatting
  --> cargo-public-api/src/error.rs:58:13
   |
58 |             write!(f, "Removed items not allowed: {:?} ", self.removed)?
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add a `;` here: `write!(f, "Removed items not allowed: {:?} ", self.removed)?;`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned

error: could not compile `cargo-public-api` due to 6 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `cargo-public-api` due to 6 previous errors
```

CI is still on rustc 1.62.1 but will soon upgrade to 1.63

Enselic reviewed (commented):
It indeed would be great if we could get this merged today, because I need to make a release tomorrow to adapt to new rustdoc JSON format ()

This looks good to me know! As usual things can be polished, but I don't want to hold back merging. To me it is very much good enough for merging. Great job! And good collab again!

So please git fetch origin, rebase on dce92ae, squash as you see fit, and let's get this merged :)

matthiasbeyer commented:
Your latest suggestions were added in this rebase+squash. I am ready to merge here! :tada:

Enselic commented:
CI falls because rustdoc JSON format changed in nightly as I mentioned 😅

I'll take care of updating rustdoc JSON and then rebase this and get it merged today 👍

matthiasbeyer commented:
Just ping me for the rebase! :wink:

Enselic commented:
Is it ok if I do it? I'll keep your commits and just add a commit to update readme and --help text

matthiasbeyer commented:
Okay, if that suits your workflow better, feel free! :tada: :+1: :smile:

Enselic commented:
👍

matthiasbeyer commented:
:tada:

Enselic commented:
Released as [v0.14.0](https://crates.io/crates/cargo-public-api/0.14.0) (:tada:)
